File: lang/typescript/core/src/middleware/iam.ts
/** IAM policy evaluation middleware. */

import type { FastifyReply, FastifyRequest } from "fastify";
import type { ServerState, IamPolicy, IamStatement } from "../types";

function wildcardMatch(pattern: string, value: string): boolean {
  const escaped = pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&");
  const regex = new RegExp(
    "^" + escaped.replace(/\*/g, ".*").replace(/\?/g, ".") + "$",
    "i"
  );
  return regex.test(value);
}

function evaluateStatement(
  stmt: IamStatement,
  action: string,
  resource: string
): "allow" | "deny" | "no_match" {
  const actions = Array.isArray(stmt.Action) ? stmt.Action : [stmt.Action];
  const resources = Array.isArray(stmt.Resource) ? stmt.Resource : [stmt.Resource];

  const actionMatch = actions.some((a) => wildcardMatch(a, action));
  const resourceMatch = resources.some((r) => wildcardMatch(r, resource));

  if (!actionMatch || !resourceMatch) return "no_match";
  return stmt.Effect === "Allow" ? "allow" : "deny";
}

function evaluatePolicy(
  policy: IamPolicy,
  action: string,
  resource: string
): "allow" | "deny" | "no_match" {
  let result: "allow" | "deny" | "no_match" = "no_match";
  for (const stmt of policy.Statement ?? []) {
    const r = evaluateStatement(stmt, action, resource);
    if (r === "deny") return "deny";
    if (r === "allow") result = "allow";
  }
  return result;
}

function isAuthorized(state: ServerState, accessKeyId: string, action: string, resource: string): boolean {
  const { iamConfig } = state;

  // Determine which identity to use
  const defaultIdentity = iamConfig.default_identity;

  // Look up identity: try exact access key, then default_identity, then "default"
  const identity =
    iamConfig.identities[accessKeyId] ??
    (defaultIdentity ? iamConfig.identities[defaultIdentity] : undefined) ??
    iamConfig.identities["default"];

  if (!identity) {
    // No identity configured — implicit deny in enforce mode (unknown callers are rejected)
    return false;
  }

  // Explicit deny check in inline policies
  for (const policy of identity.inline_policies ?? []) {
    for (const stmt of policy.Statement ?? []) {
      if (stmt.Effect === "Deny") {
        const r = evaluateStatement(stmt, action, resource);
        if (r === "deny") return false;
      }
    }
  }

  // Permission boundary check (if set, must have explicit allow)
  if (identity.permission_boundary) {
    const boundaryResult = evaluatePolicy(identity.permission_boundary, action, resource);
    if (boundaryResult !== "allow") return false;
  }

  // Identity policy allow check
  for (const policy of identity.inline_policies ?? []) {
    const r = evaluatePolicy(policy, action, resource);
    if (r === "allow") return true;
  }

  // Implicit deny
  return false;
}

export async function applyIamAuth(
  state: ServerState,
  service: string,
  operation: string,
  req: FastifyRequest,
  reply: FastifyReply,
  xmlProtocol = false
): Promise<boolean> {
  if (!state.iamConfig.enforce) return false;

  // Extract access key from Authorization header
  const authHeader = req.headers["authorization"] ?? "";
  const match = /Credential=([^/,]+)/.exec(authHeader);
  const accessKeyId = match ? match[1] : "anonymous";

  // Build action: service:Operation (e.g. states:StartExecution)
  const action = `${service.toLowerCase()}:${operation}`;
  const resource = "*";

  if (!isAuthorized(state, accessKeyId, action, resource)) {
    if (xmlProtocol) {
      reply.status(403)
        .header("Content-Type", "application/xml")
        .send(`<?xml version="1.0" encoding="UTF-8"?><Error><Code>AccessDenied</Code><Message>Access Denied: User is not authorized to perform: ${action}</Message></Error>`);
    } else {
      reply.status(403).send({
        __type: "AccessDeniedException",
        message: `User: arn:aws:iam::000000000000:user/${accessKeyId} is not authorized to perform: ${action}`,
      });
    }
    return true;
  }

  return false;
}

File: lang/go/core/lws/state/iam.go
package state

import (
	"fmt"
	"net/http"
	"regexp"
	"strings"
)

func wildcardMatch(pattern, value string) bool {
	escaped := regexp.QuoteMeta(pattern)
	escaped = strings.ReplaceAll(escaped, `\*`, ".*")
	escaped = strings.ReplaceAll(escaped, `\?`, ".")
	re, err := regexp.Compile("(?i)^" + escaped + "$")
	if err != nil {
		return false
	}
	return re.MatchString(value)
}

func toStringSlice(v interface{}) []string {
	if v == nil {
		return nil
	}
	switch t := v.(type) {
	case string:
		return []string{t}
	case []string:
		return t
	case []interface{}:
		var result []string
		for _, item := range t {
			if s, ok := item.(string); ok {
				result = append(result, s)
			}
		}
		return result
	}
	return nil
}

type evalResult int

const (
	evalNoMatch evalResult = iota
	evalAllow
	evalDeny
)

func evaluateStatement(stmt IamStatement, action, resource string) evalResult {
	actions := toStringSlice(stmt.Action)
	resources := toStringSlice(stmt.Resource)

	actionMatch := false
	for _, a := range actions {
		if wildcardMatch(a, action) {
			actionMatch = true
			break
		}
	}
	if !actionMatch {
		return evalNoMatch
	}

	resourceMatch := false
	for _, r := range resources {
		if wildcardMatch(r, resource) {
			resourceMatch = true
			break
		}
	}
	if !resourceMatch {
		return evalNoMatch
	}

	if stmt.Effect == "Allow" {
		return evalAllow
	}
	return evalDeny
}

func evaluatePolicy(policy IamPolicy, action, resource string) evalResult {
	result := evalNoMatch
	for _, stmt := range policy.Statement {
		r := evaluateStatement(stmt, action, resource)
		if r == evalDeny {
			return evalDeny
		}
		if r == evalAllow {
			result = evalAllow
		}
	}
	return result
}

func isAuthorized(cfg IamConfig, accessKeyID, action, resource string) bool {
	identity, ok := cfg.Identities[accessKeyID]
	if !ok {
		if cfg.DefaultIdentity != "" {
			identity, ok = cfg.Identities[cfg.DefaultIdentity]
		}
		if !ok {
			identity, ok = cfg.Identities["default"]
		}
		if !ok {
			return false
		}
	}

	for _, policy := range identity.InlinePolicies {
		for _, stmt := range policy.Statement {
			if stmt.Effect == "Deny" {
				if evaluateStatement(stmt, action, resource) == evalDeny {
					return false
				}
			}
		}
	}

	if identity.PermissionBoundary != nil {
		if evaluatePolicy(*identity.PermissionBoundary, action, resource) != evalAllow {
			return false
		}
	}

	for _, policy := range identity.InlinePolicies {
		if evaluatePolicy(policy, action, resource) == evalAllow {
			return true
		}
	}

	return false
}

// ApplyIAMAuth checks IAM rules. Returns true if response was sent (access denied).
func ApplyIAMAuth(s *ServerState, service, operation string, r *http.Request, w http.ResponseWriter, isXML bool) bool {
	cfg := s.GetIamConfig()
	if !cfg.Enforce {
		return false
	}

	authHeader := r.Header.Get("Authorization")
	accessKeyID := "anonymous"
	if authHeader != "" {
		re := regexp.MustCompile(`Credential=([^/,]+)`)
		if m := re.FindStringSubmatch(authHeader); len(m) > 1 {
			accessKeyID = m[1]
		}
	}

	action := fmt.Sprintf("%s:%s", strings.ToLower(service), operation)
	resource := "*"

	if !isAuthorized(cfg, accessKeyID, action, resource) {
		if isXML {
			w.Header().Set("Content-Type", "application/xml")
			w.WriteHeader(403)
			fmt.Fprintf(w, `<?xml version="1.0" encoding="UTF-8"?><Error><Code>AccessDenied</Code><Message>Access Denied: User is not authorized to perform: %s</Message></Error>`, action)
		} else {
			WriteJSONError(w, "AccessDeniedException",
				fmt.Sprintf("User: arn:aws:iam::000000000000:user/%s is not authorized to perform: %s", accessKeyID, action),
				403)
		}
		return true
	}
	return false
}
